CeedlingCompletions.py

```diff
@@ -0,0 +1,407 @@
+import re
+import sublime
+import sublime_plugin
+
+
+class CeedlingFlags:
+    Pending = 1
+    NoMatch = 2
+
+
+TEST_FUNCTIONS = [
+    [
+        "test\ttemplate test",
+        "void test_${1:function}_should_${2:behavior}(void)\n{\n\t${3}\n}\n",
+    ],
+    [
+        "testf\ttemplate test fail",
+        'void test_${1:function}_should_${2:behavior}(void)\n{\n\tTEST_FAIL_MESSAGE ("${3:Implement me!}");\n}\n',
+    ],
+    [
+        "testi\ttemplate test ignore",
+        'void test_${1:function}_should_${2:behavior}(void)\n{\n\tTEST_IGNORE_MESSAGE ("${3:Implement me!}");\n}\n',
+    ],
+]
+
+
+# Base tyoes used
+TYPES_INTEGER = {
+    "INT": ["", "8", "16", "32", "64"],
+    "UINT": ["", "8", "16", "32", "64"],
+    "size_t": [""],
+    "HEX": ["", "8", "16", "32", "64"],
+    "CHAR": [""],
+}
+
+TYPES_NUMERIC = {
+    "FLOAT": [""],
+    "DOUBLE": [""],
+}
+
+TYPES_ARRAY = {
+    "PTR": [""],
+    "STRING": [""],
+    "MEMORY": [""],
+}
+
+TYPES_STRUCT_STRING = {
+    "STRING_LEN": [""],
+}
+
+ASSERTIONS_BASIC = [
+    {"cmp3": "FAIL"},
+    {"cmp3": "IGNORE"},
+    {"cmp3": "ONLY"},
+    {"cmp3": "PASS"},
+]
+
+ASSERTIONS_BOOL = [
+    {"cmp3": "TRUE"},
+    {"cmp3": "UNLESS"},
+    {"cmp3": "FALSE"},
+]
+
+ASSERTIONS_PTR = [
+    {
+        "cmp2": "NULL",
+    },
+    {
+        "cmp2": "NOT",
+        "cmp3": "NULL",
+    },
+    {
+        "cmp2": "EMPTY",
+    },
+    {
+        "cmp2": "NOT",
+        "cmp3": "EMPTY",
+    },
+]
+
+CEEDLING_ASSERT_COMMON = [
+    {
+        "cmp1": "EQUAL",
+        "extra": ["numeric", "array", "struct"],
+    },
+    {
+        "cmp1": "NOT",
+        "cmp2": "EQUAL",
+    },
+    {
+        "cmp1": "GREATER",
+        "cmp2": "THAN",
+    },
+    {
+        "cmp1": "LESS",
+        "cmp2": "THAN",
+    },
+    {
+        "cmp1": "GREATER",
+        "cmp2": "OR",
+        "cmp3": "EQUAL",
+    },
+    {
+        "cmp1": "LESS",
+        "cmp2": "OR",
+        "cmp3": "EQUAL",
+    },
+    {
+        "p5": "WITHIN",
+        "extra": ["numeric"],
+    },
+    {
+        "p4": "ARRAY",
+        "p5": "WITHIN",
+    },
+    {
+        "cmp1": "EQUAL",
+        "p4": "ARRAY",
+        "extra": ["array", "numeric"],
+    },
+    {
+        "cmp1": "EACH",
+        "cmp2": "EQUAL",
+        "extra": ["array", "numeric"],
+    },
+]
+
+
+class CeedlingCompletions(sublime_plugin.EventListener):
+
+    pattern = None
+
+    def __init__(self):
+
+        if sublime.version().startswith("4"):
+            self.flags = (
+                sublime.DYNAMIC_COMPLETIONS | sublime.INHIBIT_WORD_COMPLETIONS
+            )
+        else:
+            self.flags = sublime.INHIBIT_WORD_COMPLETIONS
```

Comment:
Just confirming whether you're aware that this blocks all completions from identifiers used in the current (and other) views unconditionally. I didn't review the rest of this huge class enough to understand whether there is code to inject or determine the valid identifiers for the current context.

Reply:
Yes.  I'm aware.
The blocking occurs if:
- the file is c or c++
- the file includes "unity.h"
- the caret is within a function body
- the input matches a regex used to verify the format of the abbreviation

Reply:
The completions are actually for Unity unit test framework which Ceedling includes. I’ll pare the completions back and look at making the Unity completions a separate  package.

Reply:
The completions are actually for Unity unit test framework so might be better to make them a seperate package rather than over complicating Ceedling package.